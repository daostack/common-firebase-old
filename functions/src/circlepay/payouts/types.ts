import { IBaseEntity } from '../../util/types';

export type PayoutStatus = 'pending';

export interface IPayoutEntity extends IBaseEntity {
  /**
   * This is the id of the payout on circle side. Cam
   * be useful for updating payout status
   */
  circlePayoutId: string;

  /**
   * The ID of the proposal, for witch the payout
   * is made
   */
  proposalId: string;

  /**
   * The destination of the payment
   */
  destination: IPayoutDestination;

  /**
   * The current status of the payout
   */
  status: PayoutStatus;

  /**
   * The amount of the payout in US dollar cents
   */
  amount: number;

  security: IPayoutSecurity[];

  /**
   * Whether the payout has been send to circle for execution
   */
  executed: boolean;

  /**
   * Whether the payout has been voided
   */
  voided: boolean;
}

/**
 * Security details about the payout, needed for the
 * execution of it
 */
export interface IPayoutSecurity {
  /**
   * The ID of the token (no unique globally, but unique in the payout scope)
   */
  id: number;

  /**
   * The token itself
   */
  token: string;

  /**
   * Whether the token has been redeemed successfully
   */
  redeemed: boolean;

  /**
   * How many times an unsuccessful redemption was made
   */
  redemptionAttempts: number;
}

/**
 * Details about the destination
 * of the payout
 */
export interface IPayoutDestination {
  /**
   * The ID of the destination in
   * the firebase
   */
  id: string;

  /**
   * The ID of the destination on
   * circle's side
   */
  circleId: string;
}