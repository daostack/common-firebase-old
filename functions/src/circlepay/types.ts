export interface ICircleCreateCardResponse {
  data: {
    id: string;

    billingDetails: ICircleBillingDetails;

    metadata: {
      email: string;
      phoneNumber: string;
    }

    expMonth: number;
    expYear: number;
    network: 'VISA' | 'MASTERCARD';
    last4: string;
    createDate: Date;
    updateDate: Date;
  }
}



export interface ICircleCreateCardPayload {
  /**
   * Unique idempotency key. This key is utilized to ensure
   * exactly-once execution of mutating requests.
   */
  idempotencyKey: string;

  /**
   * Unique identifier of the public key used in encryption.
   */
  keyId: string;

  /**
   * PGP encrypted json string. The object format given here needs to
   * be stringified and PGP encrypted before it is sent to the server,
   * so encryptedData will end up as a string, rather than an object.
   */
  encryptedData: string;

  billingDetails: ICircleBillingDetails

  /**
   * Two digit number representing the card's expiration month.
   */
  expMonth: number;

  /**
   * Four digit number representing the card's expiration year.
   */
  expYear: number;


  metadata: {
    /**
     * Email of the user
     */
    email: string;

    /**
     * Hash of the session identifier; typically of the end user.
     */
    sessionId: string;

    /**
     * Single IPv4 or IPv6 address of user
     */
    ipAddress: string;
  }
}

interface ICircleBillingDetails {
  /**
   * Full name of the card or bank account holder.
   */
  name: string;

  /**
   * City portion of the address.
   */
  city: string;

  /**
   * Country portion of the address. Formatted as a
   * two-letter country code specified in ISO 3166-1 alpha-2.
   */
  country: string;

  /**
   * Line one of the street address.
   */
  line1: string;

  /**
   * Line two of the street address. Optional
   */
  line2: string;

  /**
   * State / County / Province / Region portion of the address. It is optional
   * except if the country is US or Canada district is required and
   * should use the two-letter code for the subdivision.
   */
  district?: string;

  /**
   * Postal / ZIP code of the address.
   */
  postalCode: string;
}