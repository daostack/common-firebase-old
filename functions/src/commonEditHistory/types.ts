import { IBaseEntity } from '../util/types';
import { ICommonUpdate, ICommonEntity } from '../common/types';

export interface ICommonEditHistory extends IBaseEntity {
	commonId: string,
	changedBy: string,
	originalDocument: ICommonEntity,
	newDocument: ICommonEntity,
}