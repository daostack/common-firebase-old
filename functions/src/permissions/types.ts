import { ICommonEntity } from '../common/types';


export interface IPermissionPayload {
	/**
	 * The common that we grant permission to
	 */
	common: ICommonEntity;// should be the new common updateable type
	
	/**
	 * The role we want to grant the user with userId
	 */
	role: Role;

	/**
	 * The ID of the user that needs to get permission
	 */
	userId: string;

	/**
	 * The ID of the user who wants to grant permission to userId
	 */
	requestByUserId?: string;
}

export interface IPermission {
	/**
	 * The role of the user
	 */
	role: Role;

	/**
	 * The commonId of the common that a user has permission to
	 */
	data?: {
		commonId?: string
	}
}

export type Role = 'founder'| 'moderator' | 'other';