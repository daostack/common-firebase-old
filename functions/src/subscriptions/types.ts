import { CirclePaymentStatus } from '../util/types';

export interface ISubscriptionEntity {
  id: string;

  cardId: string;
  userId: string;
  proposalId: string;

  payments: ISubscriptionPayment[];

  dueDate: Date;

  status: SubscriptionStatus;

  amount: number;

  metadata: ISubscriptionMetadata;

  paymentFailures?: ISubscriptionPayment[];

  revoked: boolean;
}

export interface ISubscriptionPayment {
  paymentId: string;
  paymentStatus: CirclePaymentStatus;
}

export interface ISubscriptionPaymentFailure {
 paymentId: string;

 paymentFailedOn: Date;
}

export interface ISubscriptionMetadata {
  common: {
    name: string;
    id: string;
  }
}


export type SubscriptionStatus = 'Active' | 'CanceledByUser' | 'CanceledByPaymentFailure' | 'PaymentFailed';