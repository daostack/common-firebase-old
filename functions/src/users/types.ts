import { IBaseEntity } from '../util/types';
import { IPermission } from '../permissions/types';

export interface IUserEntity extends IBaseEntity {
   uid: string;s

   email: string;
   photoURL: string;

   firstName: string;
   lastName: string;
   displayName: string;
   roles?: IPermission[];
}