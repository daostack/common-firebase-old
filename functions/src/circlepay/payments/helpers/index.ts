export { isFinalized, isSuccessful, isFailed } from './statusHelper';

export { feesHelper } from './feesHelper';
export { failureHelper } from './failureHelper';