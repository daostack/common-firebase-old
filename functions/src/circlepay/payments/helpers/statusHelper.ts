import { IPaymentEntity } from '../types';

const successfulStatuses = ['confirmed', 'paid'];
const terminalStatuses = ['paid', 'failed'];

export const isSuccessful = (payment: IPaymentEntity): boolean =>
  successfulStatuses.some(status => status === payment.status);

export const isFinalized = (payment: IPaymentEntity): boolean =>
  terminalStatuses.some(status => status === payment.status);