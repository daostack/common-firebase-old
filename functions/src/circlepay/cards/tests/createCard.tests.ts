import { v4 } from 'uuid';
import firebaseFunctionsTests from 'firebase-functions-test';

import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

// Mockers
import '../../../util/tests/helpers/mockers/getUserMocker';

import { circleApp, getTestAuthenticationToken } from '../../../util/tests/helpers';

import { ICircleCreateCardResponse } from '../../types';
import { IGetCircleCardResponse } from '../../client/getCardFromCircle';

import { createCardSuccessfulRequest, getCardSuccessfulCvvCheck } from './data/testCardCirlceResponses';
import { validCreateCardRequest } from './data/testCardRequests';

const circleMatcher = /.{1,}circle\.com\/.*/;

const test = firebaseFunctionsTests({
  projectId: 'common-tests'
});

describe('Card creation process', () => {
  afterAll(async () => {
    await test.cleanup();
  });

  it('should work', () => {
    expect(true).toBeTruthy();
  });

  it('should be healthy', async () => {
    const authToken = await getTestAuthenticationToken(v4());

    const response = await circleApp
      .get('/health')
      .set({
        Authorization: authToken
      });


    expect(response.status).toBe(200);
    expect(response.body.healthy).toBeTruthy();
  });

  describe('Card creation validation', () => {
    it('should have working mock ', async () => {
      const cardOwnerId = v4();
      const mock = new MockAdapter(axios);

      mock.onPost(new RegExp(circleMatcher.source + (/\/cards/).source))
        .reply<ICircleCreateCardResponse>(200, createCardSuccessfulRequest);

      mock.onGet(new RegExp(circleMatcher.source + (/\/cards/).source))
        .reply<IGetCircleCardResponse>(200, getCardSuccessfulCvvCheck);

      const response = await circleApp
        .post('/create-card')
        .send(validCreateCardRequest)
        .set({
          Authorization: getTestAuthenticationToken(cardOwnerId)
        });

      expect(response.status).toBe(200);

      console.log(response.body);
    });
  });
});