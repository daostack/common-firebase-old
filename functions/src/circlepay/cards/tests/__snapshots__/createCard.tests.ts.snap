// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card creation process Card creation validation should fail if the card expirary  1`] = `
Array [
  Object {
    "field": "keyId",
    "message": "keyId is a required field",
  },
]
`;

exports[`Card creation process Card creation validation should fail if the encryption key ID was not provided 1`] = `
Array [
  Object {
    "field": "keyId",
    "message": "keyId is a required field",
  },
]
`;
